api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,224 @@
+	//
+	// If clusterSelector is also set, then the selector as a whole selects
+	// Machines matching selector belonging to Clusters selected by clusterSelector.
+	// If clusterSelector is not set, it selects all Machines matching selector in
+	// all Clusters.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// clusterSelector is a label selector which selects Machines by the labels of
+	// their Clusters.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects Machines of all Clusters.
+	//
+	// If selector is also set, then the selector as a whole selects
+	// Machines matching selector belonging to Clusters selected by clusterSelector.
+	// If selector is not set, it selects all Machines belonging to Clusters
+	// selected by clusterSelector.
+	// +optional
+	ClusterSelector *metav1.LabelSelector `json:"clusterSelector,omitempty"`
+}
+
+// MachineDrainRulePodSelector defines to which Pods this MachineDrainRule should be applied.
+// +kubebuilder:validation:MinProperties=1
+type MachineDrainRulePodSelector struct {
+	// selector is a label selector which selects Pods by their labels.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects all Pods.
+	//
+	// If namespaceSelector is also set, then the selector as a whole selects
+	// Pods matching selector in Namespaces selected by namespaceSelector.
+	// If namespaceSelector is not set, it selects all Pods matching selector in
+	// all Namespaces.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// namespaceSelector is a label selector which selects Pods by the labels of
+	// their Namespaces.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects Pods of all Namespaces.
+	//
+	// If selector is also set, then the selector as a whole selects
+	// Pods matching selector in Namespaces selected by namespaceSelector.
+	// If selector is not set, it selects all Pods in Namespaces selected by
+	// namespaceSelector.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machinedrainrules,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Behavior",type="string",JSONPath=".spec.drain.behavior",description="Drain behavior"
+// +kubebuilder:printcolumn:name="Order",type="string",JSONPath=".spec.drain.order",description="Drain order"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description="Time duration since creation of the MachineDrainRule"
+
+// MachineDrainRule is the Schema for the MachineDrainRule API.
+type MachineDrainRule struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +required
+	metav1.ObjectMeta `json:"metadata"`
+
+	// +required
+	Spec MachineDrainRuleSpec `json:"spec"`
```

Comment:
Should we have a godoc on this to be complete? And Items?

Reply:
If we think that's helpful to anyone

Reply:
Just thinking about a potential future linter we could add to make sure every field has a godoc, this would be an omission

Reply:
I'll add it because I don't really mind adding it. But I assume the same could apply to every single `Items`, `TypeMeta`, `ObjectMeta` field on every single CRD that we have. Given that adding godoc comments to self-explanatory fields like that doesn't sound useful I would probably object applying a linter like this broadly (once we get to that discussion :))

Reply:
Added comments to Spec & Items